core/near-network/src/lib.rs

```diff
@@ -0,0 +1,400 @@
+use ::primitives::types::AccountId;
+use ::primitives::hash::CryptoHash;
+use ::primitives::traits::{Encode, Decode};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::net::SocketAddr;
+use std::io::{Error, ErrorKind};
+use ::tokio::net::{TcpStream, TcpListener, tcp::ConnectFuture};
+use ::tokio_codec::{Framed};
+use ::tokio::prelude::stream::SplitStream;
+use ::futures::sync::mpsc::{channel, Sender};
+use ::futures::{Poll, Stream, Async, Future, Sink};
+use ::bytes::Bytes;
+use ::log::{debug, error};
+use ::serde_derive::{Serialize, Deserialize};
+use ::rand::{thread_rng, seq::IteratorRandom};
+
+mod codec;
+use crate::codec::Codec;
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug, Serialize, Deserialize)]
+pub struct Peer {
+    /// address of peer
+    addr: SocketAddr,
+    /// peer id
+    id: PeerId,
+    /// account id, peer may not always have one
+    account_id: Option<AccountId>,
+}
+
+impl Peer {
+    pub fn new(addr: SocketAddr, id: PeerId, account_id: Option<AccountId>) -> Self {
+        Peer {
+            addr,
+            id,
+            account_id
+        }
+    }
+}
+
+/// Handshake info we exchange when first establish connection with a peer. 
+// TODO: make it secure
+#[derive(Serialize, Deserialize)]
+pub struct HandShake {
+    peer_id: PeerId,
+    account_id: Option<AccountId>,
+}
+
+/// unique identifier for nodes on the network
+// Use hash for now
+pub type PeerId = CryptoHash;
+
+// This is very awkward because we cannot copy sink and have to use a channel
+// to tunnel the data through. Definitely need to revisit this later to check for
+// better solutions
+type NearStream = (Sender<Bytes>, SplitStream<Framed<TcpStream, Codec>>);
```

Comment:
The point of writing codec is to be able to convert a stream/sink of bytes into a stream/sink of objects of our own type. The codec in `codec.rs` converts stream/sink of bytes into stream/sink of bytes, which does not achieve much.

As the result most of the code here operates with streams of bytes, which is what Codec+FramedStream meant to prevent. Instead of having this happening here https://github.com/nearprotocol/nearcore/pull/440/files#diff-0e9eae49ad6578692ac36809bfab3b5fR237 and here  https://github.com/nearprotocol/nearcore/pull/440/files#diff-0e9eae49ad6578692ac36809bfab3b5fR167 it should be happening in the codec itself.

This will also make the code of the codec very small.

Reply:
Good point. Fixed in 9896565